### ISSUE TITLE
smooth bandpass solution in frequency

### ISSUE THREAD

**Issue by paoloserra** [caracal-pipeline/caracal#1283]
Smoothing along the frequency axis can increase the SNR of the bandpass solution. This should be done with care not to smooth over real bandpass features.

As discussed at a telecon, it would be good to add a smoothing option in our `crosscal` worker. The consensus was that we would need to write a simple casacore/python module to do this. This would allow us to experiment with different smoothing algorithms. At the moment I think that a running median filter would be the best.

This could be presented to users as a new parameter along the lines of `crosscal: primary: b_freqsmooth: <nr chans>` , with default 1 (no smoothing).

**Comment by o-smirnov** [comment]
Suggestion from @LandmanBester: https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.UnivariateSpline.html, with e.g. k=3. Increasing s means more smoothing.

**Comment by landmanbester** [comment]
The advantage of using this over something like a median filter is that you can pass in weights to incorporate uncertainties in the gain solutions (eg. ignore flagged data). I don't think that is possible with the median filter in scipy but I haven't used that extensively. It will probably be a bit slower than the filters but maybe that's ok since it's a 1D problem

**Comment by paoloserra** [comment]
thanks for the useful input @landmanbester

**Comment by paoloserra** [comment]
@PeterKamphuis I've run a test and can confirm that setting, e.g., `solint="inf,4"` in CASA averages channels by a factor of 4 before solving, so the bandpass has steps of width = 4 channels. At the telecon I said that this is not good because we don't want to apply a step-function bandpass to the data. However, CASA does allow us to smooth over the steps by setting appropriate interpolation rules. So maybe this is still worth looking into.

According to the CASA [documentation](https://casa.nrao.edu/casadocs/casa-6.1.0/calibration-and-visibility-data/cal-library-syntax) the available options for frequency interpolation are `'nearest', 'linear', 'cubic', 'spline'`.

**Comment by PeterKamphuis** [comment]
@paoloserra I never noticed the step function but that might be related to the GMRT bandpass and I think I would set BPOLY for the solve. One thing though, I did always specify the unit for the call, i.e.  `solint="inf,4ch"` . The documentation is rather unclear on what is the default. I agree the logical default is channel but you know CASA.

I'll try to dig out an observation where I actually used this and look specifically for settings and the step.

**Comment by paoloserra** [comment]
> I never noticed the step function 

I don't think you would see it with visual inspection. You just get one bandpass "point" per 4 channels in this example. It would become a step function if you applied that bandpass to the full-res data with "nearest" interpolation in frequency.

**Comment by PeterKamphuis** [comment]
Well the GMRT data is 512 channels so I think that would still be noticeable as I used 5ch.

**Comment by PeterKamphuis** [comment]
I found the data set and you are right, there are only 100 solutions in the table. I didn't notice it because the noise on the channels is far greater than the step from solution to solution. We could look at how to apply it but I suspect just averaging the solutions afterwards is easier and more straightforward?

**Comment by paoloserra** [comment]
Thanks for the confirmation @PeterKamphuis .

Writing our own code to average the solutions shouldn't be too difficult. However, now that I've been reminded that there are interpolation options I'm not so sure that this is necessary. Let's think about this.

**Comment by paoloserra** [comment]
OK, a first go at it is at https://github.com/caracal-pipeline/caracal/tree/what-the-flag

This is implementing a running mean filter on real and imaginary parts of the bandpass independently. The width of. the running mean filter is given by `primary: b_smoothwindow` and `secondary: b_smoothwindow` in the `crosscal` worker. The default is `b_smoothwindow: 1` = no smoothing.

I've also tried with a running median filter but it does not seem as good. Still have to try the scipy's spline.

Here's an example:

**no smoothing**
![ps-gps2-1gc1_primary B0](https://user-images.githubusercontent.com/6591265/99727310-0f102f00-2ab8-11eb-914b-a15298ce7e81.png)

**mean smoothing, width = 9 channels**
![ps-gps2-1gc1_primary B0](https://user-images.githubusercontent.com/6591265/99727329-133c4c80-2ab8-11eb-98ce-821180ffab6e.png)

**median smoothing, width = 9 channels**
![ps-gps2-1gc1_primary B0](https://user-images.githubusercontent.com/6591265/99727341-17686a00-2ab8-11eb-8ce6-b7da8e14b746.png)
